README
stubgen - a member function stub generator for C++
http://www.radwin.org/michael/2011/05/10/stubgen/

Copyright (c) 2013 Michael J. Radwin


Description
-----------
Welcome to stubgen version 2.1.0.

stubgen is a C++ development tool that keeps code files in sync with
their associated headers.  When it finds a member function declaration
in a header file that doesn't have a corresponding implementation, it
creates an empty skeleton with descriptive comment headers.  stubgen has
several options, but see the "Brief Example" section below for an idea
of what it can do.

Use of the following software is subject to the BSD License. See COPYING
for details.


Brief Example
-------------
Suppose you have the following header file Point.h:
 
          class Point {
          public:
              Point(int x, int y);
              void addTo(const Point& other);
 
              int xValue, yValue;
          };
 
Running "stubgen -s Point.h" would produce the following file:
 
          /***********************************************
           * AUTHOR: Michael J. Radwin <mjr@acm.org>
           *   FILE: Point.cpp
           *   DATE: Mon Apr 20 17:39:05 1998
           *  DESCR:
           ***********************************************/
          #include "Point.h"
 
          /*
           *  Method: Point::Point()
           *   Descr:
           */
          Point::Point(int x, int y)
          {
          }
 
          /*
           *  Method: Point::addTo()
           *   Descr:
           */
          void
          Point::addTo(const Point& other)
          {
          }


Supported Platforms
-------------------
I've successfully built stubgen on the following platforms, using the
GNU tools make, gcc, bison, and flex:

    SPARC Solaris (2.5 and 2.6)
    SunOS 4.1.3
    SGI IRIX 5.3
    RS6000 AIX 3.2
    MS Windows NT 4.0 (using GNU bison/flex and MSVC++ 5.0)


Files
-----
The following files are included in this distribution:

    COPYING        -  the BSD License
    ChangeLog      -  a listing of changes made on various versions.
    Makefile       -  a makefile for building stubgen on unix
    README         -  this file
    etc/           -  debugging routines for use with the -d option
    lexer.l        -  flex source, generates tokens
    main.c         -  code generation routines
    parser.y       -  yacc source, parses header and code files
    stubgen.1      -  nroff-able man page
    table.[ch]     -  data structures used in parsing
    util.[ch]      -  utilities, logging routines, used in parsing
    test/          -  test header files for stubgen
    win32/         -  MS Visual C++ 5.0 project (Win32 console app)


Building
--------
1. ensure that your system has 'flex' installed.
2. ensure that your system has 'yacc', 'byacc', or 'bison' installed.
3. edit 'Makefile' and pick your favorite CC, LEX, YACC, LFLAGS and CFLAGS
4. run 'make'

We assume that flex, bison, and cc/gcc are in your path.

For win32, there's no need to build.  You'll find stubgen.exe in the
directory .\win32.


Installing
----------
There is no fancy installation procedure.  Since the binary doesn't
depend on any other files, copy it to /usr/local/bin or wherever you
choose.  You may wish to copy stubgen.1 to /usr/local/man/man1 or some
similar directory.


Acknowledgments
----------------
stubgen borrows code from:

Jutta Degener's 1995 ANSI C grammar (based on Jeff Lee's 1985
implementation):
ftp://ftp.uu.net/usenet/net.sources/ansi.c.grammar.Z
http://www.lysator.liu.se/c/ANSI-C-grammar-l.html
http://www.lysator.liu.se/c/ANSI-C-grammar-y.html

Graham D. Parrington's Stub Generator for the Arjuna project at the
University of Newcastle upon Tyne:
http://arjuna.ncl.ac.uk/

Raphael Assenat <raph@raphnet.net> contributed the -N and -l flags.

Dushara <nidujay@gmail.com> contributed the -p and -I flags to make it
easier to make stubs for 3rd party tools (e.g. #include <wx/string.h>)



ChangeLog
2013-12-18

	* Enabled non-member function stub generation. This allows the
	generation of stubs for C/non-member C++ functions.

	* fixed: crash on typedef of unnamed struct.

        typedef unnamed struct declarations were trying to free unallocated
        memory. In fixing this it was discovered that the unnamed structs
        were generating non-member functions with the unnamed struct as the
        name. This bug is also fixed in this commit.

2011-05-11

	* Dushara <nidujay@gmail.com> contributed the -p and -I flags to
	make it easier to make stubs for 3rd party tools (e.g. #include
	<wx/string.h>)

2004-05-12

	* Added -N and -l flags courtesy Raphael Assenat
	<raph@raphnet.net>

	* No longer bundling getopt() and basename() implementations since
	these are standard in modern libc.

	* Switched to BSD License

1998-09-14

	* main.c: Version 2.05

1998-07-27

	* lexer.l: added some c++ keywords
	need to handle typename, using, and namespace

1998-07-06

	* main.c, parser.y:
	removed extra space from throw_decl, cleaned up memory leak
	in ctor skeleton

1998-06-12

	* main.c: added support for .hh

1998-06-11

	* stubgen.1, README: Updated descriptive text

	* main.c, stubgen.1: changed standard extension to .cpp (previously .C)
	and added .hpp/.hxx support for header files

	* main.c: changed revision to thousands-based integer

	* main.c, parser.y, stubgen.1, table.c, table.h:
	allow for empty class declarations, such as
	class Element {};
	also, differentiate structs from classes with
	new STRUCT_KIND tag.
	New version: 2.04

1998-05-27

	* pathname.c: made strip_trailing_slashes static

	* main.c: removed revision log from comment header

	* README, main.c, stubgen.1: radwin.org

	* pathname.h, table.h, util.c, util.h:
	adorned header files with __cplusplus macro

1998-05-11

	* README: Bumped build number.

	* README, Makefile, lexer.l, main.c, parser.y, pathname.c, pathname.h, stubgen.1, table.c, table.h, util.c, util.h:
	Version 2.03 (updated copyright information).

	* stubgen.1: method header "styles" vs. "options"

	* stubgen.1: changed the bulleted lists and added web page

1998-04-21

	* stubgen.1:
	fixed RCS id string by blocking out with useless nroff macros

1998-04-20

	* stubgen.1: added lots of examples

1998-04-07

	* README: added supported platforms section

	* README: bumped version number, added build number, included
	ChangeLog in file listing.

	* parser.y: changed error-handling code significantly.
	several functions now return a value, and instead of
	calling fatal(), we do a YYABORT or YYERROR to get out
	of the parsing state.
	New version: 2.02.

	* main.c: removed YYDEBUG, and useless reference to yytext.
	setting tokens_seen state to 0.
	New version: 2.02.

	* table.c: made warning go away regarding malloc() cast.

	* util.c, util.h:
	made some warnings go away, removed extra yyerror() prototype.

	* lexer.l: changed error-handling code significantly.  functions
	like count() are now contributing to linebuf so we get correct
	parse error messages during lineno == 1 and other situations.
	also, instead of calling fatal() for collect*() functions,
	we return -1 and let the parser recover more gracefully.

	* Makefile: removed YYDEBUG

	* table.c: misc. changes

1998-03-27

	* main.c, parser.y:
	working on slightly better error recovery; not done yet.

	* stubgen.1: misc. changes

	* main.c, parser.y, table.c, table.h:
	added multi-line function parameters
	also changed pointer and reference (* and &) types so there
	is no trailing space before the parameter name.

1998-03-09

	* stubgen.1: added -g option information

	* main.c: misc. changes

1998-02-27

	* Makefile: clean rule


test/Rational.H
/****************************************************************
 *    NAME: Michael J. Radwin
 *    ACCT: mradwin
 *    FILE: Rational.H
 *    DATE: Sat Mar  4 01:47:16 EST 1995
 *     $Id: Rational.H,v 1.4 1998/03/13 19:18:42 mradwin Exp $
 ****************************************************************/

class Rational {
  friend ostream & operator<<(ostream &output, const Rational &r);
  friend Rational pow(const Rational &r, int exp);
  friend Rational operator+(int n, const Rational &r);
  friend Rational operator-(int n, const Rational &r);
  friend Rational operator*(int n, const Rational &r);
  friend Rational operator/(int n, const Rational &r);
  friend int operator==(int n, const Rational &r);
  friend int operator!=(int n, const Rational &r);
  friend int operator< (int n, const Rational &r);
  friend int operator> (int n, const Rational &r);
  friend int operator<=(int n, const Rational &r);
  friend int operator>=(int n, const Rational &r);

public:
  Rational(int num, int den = 1);
  Rational(const Rational &r);
  ~Rational();

  Rational operator+(const Rational &r) const;
  Rational operator-(const Rational &r) const;
  Rational operator*(const Rational &r) const;
  Rational operator/(const Rational &r) const;
  const Rational & operator= (const Rational &r);
  const Rational & operator+=(const Rational &r);
  const Rational & operator-=(const Rational &r);
  const Rational & operator*=(const Rational &r);
  const Rational & operator/=(const Rational &r);

  Rational operator+(int r) const;
  Rational operator-(int r) const;
  Rational operator*(int r) const;
  Rational operator/(int r) const;
  const Rational & operator= (int r);
  const Rational & operator+=(int r);
  const Rational & operator-=(int r);
  const Rational & operator*=(int r);
  const Rational & operator/=(int r);

  Rational operator-() const;
  int operator! () const;
  int operator==(const Rational &r) const;
  int operator==(int r) const;
  int operator!=(const Rational &r) const;
  int operator!=(int r) const;
  int operator< (const Rational &r) const;
  int operator< (int r) const;
  int operator> (const Rational &r) const;
  int operator> (int r) const;
  int operator<=(const Rational &r) const;
  int operator<=(int r) const;
  int operator>=(const Rational &r) const;
  int operator>=(int r) const;

private:
  Rational() {}
  static int compute_gcd(int a, int b);
  static int gcd(int bigger, int smaller);
  int n, d;
};

// global prototypes
ostream & operator<<(ostream &output, const Rational &r);
Rational pow(const Rational &r, int exp);
Rational operator+(int n, const Rational &r);
Rational operator-(int n, const Rational &r);
Rational operator*(int n, const Rational &r);
Rational operator/(int n, const Rational &r);
int operator==(int n, const Rational &r);
int operator!=(int n, const Rational &r);
int operator< (int n, const Rational &r);
int operator> (int n, const Rational &r);
int operator<=(int n, const Rational &r);
int operator>=(int n, const Rational &r);


test/non_member_fn.h
#ifndef CHEADER_H
#define CHEADER_H

extern int some_c_func(const void* param);

#endif


test/typedef_struct.h
#ifndef TYPEDEF_STRUCT_H
#define TYPEDEF_STRUCT_H

typedef struct {
    int id;
    int count;
} record_t;

#endif


main.c
/*
 *  FILE: main.c
 *  AUTH: Michael J. Radwin <michael@radwin.org>
 *
 *  DESC: stubgen code generation routines
 *
 *  DATE: Thu Nov 13 13:28:23 PST 1997
 *
 *  Copyright (c) 2004  Michael J. Radwin
 *
 *  Redistribution and use in source and binary forms, with or
 *  without modification, are permitted provided that the following
 *  conditions are met:
 *
 *  - Redistributions of source code must retain the above
 *    copyright notice, this list of conditions and the following
 *    disclaimer.
 *
 *  - Redistributions in binary form must reproduce the above
 *    copyright notice, this list of conditions and the following
 *    disclaimer in the documentation and/or other materials
 *    provided with the distribution. 
 *
 *  - Neither the name of stubgen nor the names of its contributors
 *    may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 *  CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 *  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 *  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 *  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 *  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 *  IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 *  THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "table.h"
#include "util.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <assert.h>
#include <unistd.h>
#include <libgen.h>

#ifdef WIN32
#include <windows.h> /* defintion of GetUserName(), BOOL, etc. */
#include <lmaccess.h>
#include <malloc.h>  /* defintion of alloca() */
#else /* !WIN32 */
#include <pwd.h>
#endif /* WIN32 */

/* protos */
static void debug_printf(syntaxelem_t *);
static void generate_skel(syntaxelem_t *);
static void generate_function_skel(syntaxelem_t *);
static void print_function(syntaxelem_t *);
static void function_hdr(syntaxelem_t *);
static void file_hdr();
static void scan_and_generate(FILE *);
static void scan_existing_skeleton();

/* duplicating variable names from stubgen.pl */
static const char *OPTS = "hqrivgae:t:p:lNIcdbfsn";
int opt_h = 0, opt_q = 0, opt_r = 0, opt_i = 0;
int opt_v = 0, opt_g = 0, opt_a = 0;
int opt_c = 0, opt_d = 0;
int opt_b = 0, opt_f = 0, opt_s = 0, opt_n = 0;
char *opt_e = "cpp";
char *opt_t = "./";
char *opt_p = "";
int opt_l = 0, opt_N = 0;
int opt_I = 0;

int using_stdio = 0;
static int new_functions = 0, fileOpened = 0, fileExisted = 0;
static int inform_indent = 0;

#ifdef SGDEBUG
static const char *logfilename = "stubgen.log";
#endif /* SGDEBUG */

FILE *outfile = NULL;
static char *inPath = NULL, *outPath = NULL;
char *currentFile = "";
static const char *lots_of_stars = 
  "***********************************************************************";
static const char *progname = "stubgen";
static const char *progver = "2.1.0";

static const char *copyright =
  "Copyright (c) 2004  Michael J. Radwin";

static const char *version_info = 
"Contributions from Raphael Assenat <raph@raphnet.net>\n\
Distributed under the BSD License.\n\
See http://www.radwin.org/michael/2011/05/10/stubgen/ for more information.\n";

static const char *usage = 
"usage: %s [-hqrivgacdlNI] [-e ext] [-t dir] [-p dir] [-{bfsn}] [infiles]\n\
  OPTIONS\n\
    -h        Display usage information.\n\
    -q        Quiet mode, no status information while generating code.\n\
    -r        Make RCS-style file headers.\n\
    -i        Don't put the #include \"my_file.H\" directive in my_file.cpp\n\
    -v        Display version information.\n\
    -g        Generate dummy return statements for functions.\n\
    -a        Split function arguments over multiple lines.\n\
    -c        Print debugging output with cerrs (#include <iostream.h>).\n\
    -d        Print debugging output with dprintfs (#include <Debug.H>).\n\
    -e ext    Generate source files with extension '.ext' (default '.cpp').\n\
    -t dir    Put the generated files in directory 'dir' (default: ./)\n\
    -p dir    Add prefix to #include directive.\n\
    -I        Use <> instead of \"\" for include directive.\n\
    -l        Put return type and function name on the same line.\n\
    -N        No file header.\n\
\n\
  METHOD HEADER STYLES\n\
    -b        Block method headers (default).\n\
    -f        Full method headers: like block, but less asterisks.\n\
    -s        Simple method headers: only \"Method\" and \"Descr\" fields.\n\
    -n        No method headers.\n";


static void generate_skel(syntaxelem_t *elt)
{
    syntaxelem_t *e;

    log_printf("generate_skel called: %s\n", elt->name);
    if (elt->kind != CLASS_KIND && elt->kind != STRUCT_KIND)
	return;

    inform_user("%*s==> %s %s", inform_indent, "", 
		(elt->kind == CLASS_KIND ? "class " : "struct"),
		elt->name);
    if (inform_indent == 0)
      inform_user(" (file %s)", inPath);
    else
      inform_user(" (nested class)");
    inform_user("\n");
    inform_indent += 4;

    for (e = elt->children; e != NULL; e = e->next) {
	if (e->kind == FUNC_KIND) {
        generate_function_skel(e);
	} else if (e->kind == CLASS_KIND || e->kind == STRUCT_KIND) {
	    /* nested class */
	    char *tmp_str = (char *) malloc(strlen(elt->name) + strlen(e->name) + 3);
	    sprintf(tmp_str, "%s::%s", elt->name, e->name);
	    free(e->name);
	    e->name = tmp_str;

	    log_printf(">>>>>>> generating NESTED %s: %s\n",
		      string_kind(e->kind), e->name);
	    print_se(e);
	    generate_skel(e);
	} else {
	    log_printf("------> ignoring   %s: %s\n",
		      string_kind(e->kind), e->name);
	}
    }

    inform_indent -= 4;
    inform_user("%*s==> %s %s", inform_indent, "",
		(elt->kind == CLASS_KIND ? "class " : "struct"),
		elt->name);
    if (inform_indent == 0)
	inform_user(" (%d functions appended to %s)", new_functions, outPath);
    else
	inform_user(" (end nested class)");
    inform_user("\n");

    elt->kind = DONE_CLASS_KIND;
}

static void generate_function_skel (syntaxelem_t * e)
{
    char *arg_str = args_to_string (e->args, 0);
    log_printf (">>>>>>> generating %s: %s %s(%s) %s\n",
                string_kind (e->kind),
                e->ret_type, e->name, arg_str,
                (e->const_flag) ? "const" : "");
    free (arg_str);
    print_se (e);
    print_function (e);
    e->kind = DONE_FUNC_KIND;
}



static void print_function(syntaxelem_t *elt)
{
    syntaxelem_t *e;

    if (find_skeleton(elt)) {
      log_printf("find_skeleton() returned true for this elt:\n");
      print_se(elt);
      return;
    }

    new_functions++;
    if (!fileOpened) {
	fileOpened = 1;

	if (using_stdio) {
	  fileExisted = 0;
	} else {
	  /* test for existence */
	  outfile = fopen(outPath, "r");
	  if (outfile != NULL) {
	    fileExisted = 1;
	    fclose(outfile);
	  }

	  /* do the fopen */
	  log_printf("writing to %s\n", outPath);
	  outfile = fopen(outPath, "a");
	  if (outfile == NULL) {
	    /* open failed */
	    fatal(1, "%s: cannot open %s\n", progname, outPath);
	  }
	}
  
	if (!fileExisted) file_hdr();
    }

    inform_user("%*s%s\n", inform_indent, "", elt->name);
    function_hdr(elt);
    for (e = elt->parent; e != NULL; e = e->parent) {
      if (e->templ) {
	fprintf(outfile, "%s\n", e->templ);
	break;
      }
    }

    {  /* scope for local vars */
	char *arg_str;

	if (opt_l) {
        if (elt->parent)
        {
            fprintf(outfile, "%s%s%s::%s(",
                elt->ret_type, (strcmp(elt->ret_type, "") ? " " : ""),
                elt->parent->name, elt->name);
        }
        else
        {
            fprintf(outfile, "%s%s%s(",
                elt->ret_type, (strcmp(elt->ret_type, "") ? " " : ""),
                elt->name);
        }
	}
	else
	{
        if (elt->parent)
        {
            fprintf(outfile, "%s%s%s::%s(",
                elt->ret_type, (strcmp(elt->ret_type, "") ? "\n" : ""),
                elt->parent->name, elt->name);
        }
        else
        {
            fprintf(outfile, "%s%s%s(",
                elt->ret_type, (strcmp(elt->ret_type, "") ? "\n" : ""),
                elt->name);
        }
	}

    if (elt->parent)
    {
        arg_str = args_to_string(
            elt->args, 
            opt_a ? strlen(elt->parent->name) + strlen(elt->name) + 3 : 0);
    }
    else
    {
        arg_str = args_to_string(
            elt->args, 
            opt_a ? strlen(elt->name) + 3 : 0);
    }
	
	fprintf(outfile, "%s)", arg_str);
	free(arg_str);

	if (elt->throw_decl)
	    fprintf(outfile, " %s", elt->throw_decl);

	if (elt->const_flag)
	    fprintf(outfile, " const");

	fprintf(outfile, "\n{\n");
    }

    debug_printf(elt);
    fprintf(outfile, "}\n\n\n");
}

static void function_hdr(syntaxelem_t *elt)
{
  if (opt_n)
    return;
  
  fprintf(outfile, "/%s\n", (opt_b ? lots_of_stars : "*"));
  
  if (elt->parent)
  {
      fprintf(outfile, " *  Method: %s::%s%s\n", elt->parent->name, elt->name,
          (opt_s ? "()" : ""));
  }
  else
  {
      fprintf(outfile, " *  Method: %s%s\n", elt->name,
          (opt_s ? "()" : ""));
  }

  if (opt_s) {
    fprintf(outfile, " *   Descr: \n");

  } else {
    char *arg_str = args_to_string(elt->args, 0);
    fprintf(outfile, " *  Params: %s\n", arg_str);
    if (strcmp(elt->ret_type, ""))
      fprintf(outfile, " * Returns: %s\n", elt->ret_type);
    fprintf(outfile, " * Effects: \n");
    free(arg_str);
  }

  fprintf(outfile, " %s/\n", (opt_b ? lots_of_stars : "*"));
}

#ifdef WIN32
static BOOL win32_fullname(const char *login, char *dest)
{
    WCHAR  wszLogin[256];              /* Unicode user name */
    struct _USER_INFO_10 *ui;          /* User structure */
    
    /* Convert ASCII user name to Unicode. */
    MultiByteToWideChar(CP_ACP, 0, login,
			strlen(login)+1, wszLogin, sizeof(wszLogin));
    
    /* Look up the user on the DC.  This function only works for
     * Windows NT, and not Windows 95. */
    if (NetUserGetInfo(NULL, (LPWSTR) &wszLogin, 10, (LPBYTE *) &ui))
	return FALSE;
    
    /* Convert the Unicode full name to ASCII. */
    WideCharToMultiByte(CP_ACP, 0, ui->usri10_full_name,
			-1, dest, 256, NULL, NULL);
    
    return TRUE;
}
#endif /* WIN32 */

static char *sg_getlogin()
{
    static char *login;
#ifdef WIN32
    static char login_buffer[256];
    DWORD size;
#endif /* WIN32 */
    static int sg_getlogin_called = 0;

    if (sg_getlogin_called)
	return login;
    else
	sg_getlogin_called = 1;

#ifdef WIN32
    if ((login = getenv("USERNAME")) == NULL) {
	if ((login = getenv("USER")) == NULL) {
	    size = 255;
	    login = login_buffer;
	    if (GetUserName(login_buffer, &size) == FALSE)
		login = "nobody";
	}
    }
#else /* !WIN32 */
    if ((login = getenv("USER")) == NULL)
	if ((login = getlogin()) == NULL)
	    login = "nobody";
#endif /* WIN32 */
    
    return login;
}

static char *sg_getfullname(const char *login)
{
    char *fullname;
    static char fullname_buffer[256];
#ifndef WIN32
    char *comma;
    struct passwd *pw;
#endif /* WIN32 */

#ifdef WIN32
    fullname = fullname_buffer;
    
    if (win32_fullname(login, fullname_buffer) == FALSE)
	fullname = "nobody";
#else /* !WIN32 */
    if ((fullname = getenv("NAME")) == NULL) {
	setpwent();
	pw = getpwnam(login);
	if (pw == NULL) {
	    fullname = "nobody";
	} else {
	    strncpy(fullname_buffer, pw->pw_gecos, 256);
	    comma = strchr(fullname_buffer, ',');
	    if (comma) *comma = '\0';
	    fullname = fullname_buffer;
	}
	endpwent();
    }
#endif /* WIN32 */

    return fullname;
}


static void file_hdr() {
  char domain[256], *tmp;
  time_t now = time(0);
  char *today = ctime(&now);
  char *login = sg_getlogin();
  char *fullname = sg_getfullname(login);
  

  log_printf("login: %s, full name: %s\n", login, fullname);

  domain[0] = '\0';
  if ((tmp = getenv("STUBGEN_DOM")) != NULL)
      sprintf(domain, "@%s", tmp);

  if (!opt_N)
  {
	  if (opt_r) {
	    fprintf(outfile, "/*\n *  FILE: %s\n *  AUTH: %s <%s%s>\n", outPath, fullname, login, domain);
	    /* no '\n' needed with ctime() */
	    fprintf(outfile, " *\n *  DESC: \n *\n *  DATE: %s *   %sId$ \n *\n", today, "$");
	    fprintf(outfile, " *  %sLog$\n", "$");
	    fprintf(outfile, " *\n */\n");
	
	  } else {
	    fprintf(outfile, "/%s\n", lots_of_stars);
	    fprintf(outfile, " * AUTHOR: %s <%s%s>\n", fullname, login, domain);
	    fprintf(outfile, " *   FILE: %s\n", outPath);
	    fprintf(outfile, " *   DATE: %s", today); /* no '\n' needed with ctime() */
	    fprintf(outfile, " *  DESCR: \n");
	    fprintf(outfile, " %s/\n", lots_of_stars);
	  }
  }

  if (!opt_i && !using_stdio) {
      if(opt_I)
        fprintf(outfile, "#include <%s%s>\n", opt_p, inPath);
      else 
        fprintf(outfile, "#include \"%s%s\"\n", opt_p, inPath);
  }
  if (opt_c)
    fprintf(outfile, "#include <iostream.h>\n");
  else if (opt_d)
    fprintf(outfile, "#include <Debug.H>\n");

  fprintf(outfile, "\n");
}
    

static void debug_printf(syntaxelem_t *elt)
{
  /*
   * Make a dummy return value if this function is not a 
   * procedure (returning void) or a ctor (returning "").
   * Don't bother for references because they require an lvalue.
   */
  int dummy_ret_val = (opt_g && 
		       strcmp(elt->ret_type, "") != 0 && 
		       strcmp(elt->ret_type, "void") != 0 &&
		       strchr(elt->ret_type, '&') == 0);
  /* 
   * If it's not a pointer type, create a temporary on the stack.
   * They'll get warnings "dummy has not yet been assigned a value"
   * from the compiler, but that's the best we can do.
   * If it's a pointer type, return NULL (we assume it's defined).
   */
  if (dummy_ret_val != 0 && strchr(elt->ret_type, '*') == NULL)
    fprintf(outfile, "    %s dummy;\n\n", elt->ret_type);

  if (opt_c)
    fprintf(outfile, "    cerr << \"%s::%s()\" << endl;\n",
	    elt->parent->name, elt->name);
  else if (opt_d)
    fprintf(outfile, "    dprintf((\"%s::%s()\\n\"));\n",
	    elt->parent->name, elt->name);

  if (dummy_ret_val != 0)
    fprintf(outfile, "    return %s;\n",
	    (strchr(elt->ret_type, '*') == NULL) ? "dummy" : "NULL");
}
    
extern char linebuf[];
extern int lineno;
extern int column;
extern int tokens_seen;
extern int yyparse();

static void scan_existing_skeleton()
{
  extern FILE *yyin;
  FILE *existing;

  log_printf("checking for existence of %s ...\n", outPath);
  if ((existing = fopen(outPath, "r")) == NULL)
    return;

  log_printf("%s exists, scanning skeleton...\n", outPath);
  inform_user("scanning %s ...\n", outPath);

  lineno = 1;
  column = 0;
  tokens_seen = 0;

  yyin = existing;
  currentFile = outPath;
  linebuf[0] = '\0';
  yyparse();
  log_printf("finished yyparse()\n");
  currentFile = "";

  fclose(existing);
  log_printf("done scanning skeleton...\n");
}


static void scan_and_generate(FILE *infile)
{
  extern FILE *yyin;
  
  fileOpened = 0;
  lineno = 1;
  column = 0;
  tokens_seen = 0;

  /* normal interaction on yyin and outfile from now on */
  inform_user("parsing %s ...\n", inPath);
  log_printf("parsing %s ...\n", inPath);
  yyin = infile;
  currentFile = inPath;
  linebuf[0] = '\0';
  yyparse();
  log_printf("finished yyparse()\n");
  currentFile = "";
  log_printf("expanding classes...\n");
  while (!class_queue_empty()) {
    syntaxelem_t *elt = dequeue_class();
    generate_skel(elt);
  }

    log_printf ("expanding non member functions...\n");
    while (!non_member_function_queue_empty ())
    {
        syntaxelem_t *elt = dequeue_non_member_function ();
        generate_function_skel (elt);
    }
    
  log_printf("closing %s\n", outPath);
  free(outPath);
  outPath = NULL;
  if (fileOpened) {
    fflush(outfile);
    fclose(outfile);
    outfile = NULL;
  }
  log_printf("done with %s\n", inPath);
}

#ifndef _MAX_PATH
#define _MAX_PATH 256
#endif /* !def _MAX_PATH */

int main(int argc, char **argv) 
{ 
    extern int optind;
    extern char *optarg;
    int c, err_flag = 0;
    char *ext;

#ifdef SGDEBUG
    char logfilename_buffer[_MAX_PATH];
#ifdef WIN32
    DWORD tmpPathLen;
#endif /* WIN32 */
#endif /* SGDEBUG */

#ifdef SGDEBUG
#ifdef WIN32
    tmpPathLen = GetTempPath(_MAX_PATH, logfilename_buffer);
    if (logfilename_buffer[tmpPathLen - 1] != '\\')
	strcat(logfilename_buffer, "\\");
#else /* !WIN32 */
    strcpy(logfilename_buffer, "/tmp/");
#endif /* WIN32 */

    strcat(logfilename_buffer, sg_getlogin());
    strcat(logfilename_buffer, "-");
    strcat(logfilename_buffer, logfilename);

    if (!log_open(logfilename_buffer)) {
      /* open failed */
      fatal(1, "%s: cannot write to %s\n", progname, logfilename_buffer);
    }
#endif /* SGDEBUG */

    while ((c = getopt(argc, argv, OPTS)) != EOF) {
	switch (c) {
	case 'h':
	    opt_h = 1; break;
	case 'q':
	    opt_q = 1; break;
	case 'v':
	    opt_v = 1; break;
	case 'g':
	    opt_g = 1; break;
	case 'a':
	    opt_a = 1; break;
	case 'e':
 	    opt_e = optarg; 
	    if (opt_e[0] == '.')
	      opt_e++;
	    break;
	case 't':
		opt_t = optarg;
		break;
	case 'p':
		opt_p = optarg;
		break;
	case 'I':
		opt_I = 1; break;
	case 'l':
		opt_l = 1; break;
	case 'N':
		opt_N = 1; break;
	case 'r':
	    opt_r = 1; break;
	case 'i':
	    opt_i = 1; break;
	case 'c':
	    if (opt_d) err_flag = 1;
	    opt_c = 1; break;
	case 'd':
	    if (opt_c) err_flag = 1;
	    opt_d = 1; break;
	case 'b':
	    if (opt_f || opt_s || opt_n) err_flag = 1;
	    opt_b = 1; break;
	case 'f':
	    if (opt_b || opt_s || opt_n) err_flag = 1;
	    opt_f = 1; break;
	case 's':
	    if (opt_f || opt_b || opt_n) err_flag = 1;
	    opt_s = 1; break;
	case 'n':
	    if (opt_f || opt_s || opt_b) err_flag = 1;
	    opt_n = 1; break;
	default:
            err_flag = 1;
	}
    }
    
    if (opt_h || opt_v || err_flag) {
      inform_user("%s version %s\n",
		  progname, progver);
      if (opt_h || err_flag)
	fatal(err_flag, usage, progname);
      else
	fatal(0, "%s\n%s", copyright, version_info);
    }

    if (!(opt_b || opt_f || opt_s || opt_n))
	opt_b = 1;

    /* done setting options */
    if (argc == optind) {
	/* read from stdin and stdout */
        outfile = stdout;
	fprintf(outfile, "/* %s: reading from stdin */\n", progname);
	using_stdio = 1;
	inPath = "stdin";
	outPath = strdup("stdout");
	
        opt_q = 1;  /* force quiet mode */
	log_printf("initting...\n");
	init_tables();
	scan_and_generate(stdin);
	log_printf("freeing memory...\n");
	free_tables();

    } else {
        inform_user("%s version %s\n",
		    progname, progver);

	/* each bloody file from the command line */
	while (optind < argc) {
  	    FILE *infile;
	    log_printf("working on %s\n", argv[optind]);
	    /* open for read */
	    infile = fopen(argv[optind], "r");
	    if (infile == NULL) {
	      /* open failed */
	      fatal(1, "%s: cannot open %s\n", progname, argv[optind]);
	    }

	    inPath = basename(argv[optind]);
		outPath = (char *)malloc(strlen(opt_t) + strlen(inPath) + strlen(opt_e) + 3); /* +3 for / . and NUL */
		strcpy(outPath, opt_t);
		strcat(outPath, "/");
		strcat(outPath, inPath);

	    /* tie off .h, .hh, .hpp, or .hxx extension */
	    if (((ext = strrchr(outPath, '.')) != NULL) &&
		(((strlen(ext) == 2) &&
		  ((ext[1] == 'H') || (ext[1] == 'h'))) ||
		 (((strlen(ext) == 3) &&
		   ((ext[1] == 'H') || (ext[1] == 'h')) &&
		   ((ext[2] == 'H') || (ext[2] == 'h')))) ||
		 ((strlen(ext) == 4) &&
		  ((ext[1] == 'H') || (ext[1] == 'h')) &&
		  ((((ext[2] == 'P') || (ext[2] == 'p')) &&
		    ((ext[3] == 'P') || (ext[3] == 'p'))) ||
		   (((ext[2] == 'X') || (ext[2] == 'x')) &&
		    ((ext[3] == 'X') || (ext[3] == 'x')))))))
		*ext = '\0';
	    
	    assert(opt_e[0] != '.');
	    strcat(outPath, ".");
	    strcat(outPath, opt_e);

	    log_printf("initting...\n");
	    init_tables();
	    scan_existing_skeleton();
	    scan_and_generate(infile);
	    log_printf("freeing memory...\n");
	    free_tables();
	    clear_skeleton_queue();
	    fclose(infile);
	    optind++;
	}
    }

#ifdef SGDEBUG
    log_flush();
    log_close();
#endif /* SGDEBUG */

    return 0;
}
